src/transform/src/normalize_lets.rs

```diff
@@ -351,121 +351,181 @@ mod let_motion {
+    /// A stand in for the types of bindings we might encounter.
     ///
-    /// The traversal is only of the `LetRec` nodes, for which fear of stack exhaustion is nominal.
-    fn post_order_harvest_lets(expr: &mut MirRelationExpr) {
-        if let MirRelationExpr::LetRec {
-            ids,
-            values,
-            limits,
-            body,
-        } = expr
-        {
-            // Only recursively descend through `LetRec` stages.
-            for value in values.iter_mut() {
-                post_order_harvest_lets(value);
+    /// As we dissolve various `Let` and `LetRec` expressions, a `Binding` will carry
+    /// the relevant information as we hoist it to the root of the expression.
+    enum Binding {
+        // Binding resulting from a `Let` expression.
+        Let(LocalId, MirRelationExpr),
+        // Bindings resulting from a `LetRec` expression.
+        LetRec(Vec<(LocalId, MirRelationExpr, Option<LetRecLimit>)>),
+    }
+
+    /// Hoist all exposed bindings to the root of the expression.
+    ///
+    /// A binding is "exposed" if the path from the root does not cross a LetRec binding.
+    /// After the call, the expression should be a linear sequence of bindings, where each
+    /// `Let` binding is of a let-free expression. There may be `LetRec` expressions in the
+    /// sequence, and their bindings will have hoisted bindings to their root, but not out
+    /// of the binding.
+    fn hoist_bindings(expr: &mut MirRelationExpr) {
+        // Bindings we have extracted but not fully processed.
+        let mut worklist = Vec::new();
+        // Bindings we have extracted and then fully processed.
+        let mut finished = Vec::new();
+
+        extract_bindings(expr, &mut worklist);
+        while let Some(mut bind) = worklist.pop() {
+            match &mut bind {
+                Binding::Let(_id, value) => {
+                    extract_bindings(value, &mut worklist);
+                }
+                Binding::LetRec(_binds) => {
+                    // nothing to do here; we cannot hoist letrec bindings and refine
+                    // them in an outer loop.
+                }
             }
+            finished.push(bind);
+        }
 
-            let mut bindings = BTreeMap::new();
-            for (id, mut value, max_iter) in
-                izip!(ids.drain(..), values.drain(..), limits.drain(..))
-            {
-                bindings.extend(harvest_non_recursive(&mut value));
-                bindings.insert(id, (value, max_iter));
+        // The worklist is empty and finished should contain only LetRec bindings and Let
+        // bindings with let-free expressions bound. We need to re-assemble them now in
+        // the correct order. The identifiers are "sequential", so we should be able to
+        // sort by them, with some care.
+
+        // We only extract non-empty letrec bindings, so it is safe to peek at the first.
+        finished.sort_by_key(|b| match b {
```

Comment:
I suspect this could be dropped and replaced by `finished.reverse()` or iterating through the reversed list in the `for` loop.